### PR TITLE
add a new `device` package for device detection

### PR DESCRIPTION
**Pull request by joeshaw**
Calling `device.Lookup()` returns a `Device` struct with accessor
methods to get information on the user agent such as name/brand/model as
well as booleans for common device types.

Based on the API in the [Rust SDK](https://docs.rs/fastly/latest/fastly/device_detection/index.html) and our VCL variables.

### CODE REVIEW

**Comment by joeshaw** [comment]
This new API requires JSON decoding, but I've tested it on Go and TinyGo, and the integration tests will ensure it works on our supported configurations.

File: internal/abi/fastly/hostcalls_guest.go [hunk 1]
```diff
@@ -3387,3 +3387,55 @@ func PurgeSurrogateKey(surrogateKey string, opts PurgeOptions) error {
...
+//go:noescape
+func fastlyDeviceDetectionLookup(
+	userAgentData prim.Pointer[prim.U8], userAgentLen prim.Usize,
+	buf prim.Pointer[prim.Char8],
+	bufLen prim.Usize,
+	nWritten prim.Pointer[prim.Usize],
+) FastlyStatus
+
+func DeviceLookup(userAgent string) ([]byte, error) {
+	buf := prim.NewWriteBuffer(defaultBufferLen)
```
Comment (fgsch) [inline internal/abi/fastly/hostcalls_guest.go:3414]: Is the default buffer length the same in other SDKs?

File: integration_tests/device_detection/fastly.toml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+
+[local_server]
+[local_server.device_detection]
+format = "inline-toml"
+
+[local_server.device_detection.user_agents]
+[local_server.device_detection.user_agents."Mozilla/5.0 (X11; Linux x86_64; rv:10.0) Gecko/20100101 Firefox/10.0 [FBAN/FBIOS;FBAV/8.0.0.28.18;FBBV/1665515;FBDV/iPhone4,1;FBMD/iPhone;FBSN/iPhone OS;FBSV/7.0.4;FBSS/2; FBCR/Telekom.de;FBID/phone;FBLC/de_DE;FBOP/5]"]
+user_agent = {}
+os = {}
```
Comment (fgsch) [inline integration_tests/device_detection/fastly.toml:19]: Does it make sense to test user_agent and os too?

File: internal/abi/fastly/hostcalls_guest.go (hunk 1 above)
Comment (joeshaw) [reply internal/abi/fastly/hostcalls_guest.go:3414]: In our SDK it's 16k.  In Rust it's only 1k.  The Rust SDK is better about using "adaptive buffers", in which you make a hostcall with a certain sized buffer and if it's too small the SDK tells you the size it needs, and you allocate a new buffer to fit it.  We do it for a few hostcalls in the Go SDK (and any new one I write, including this one) but it isn't pervasive.

I can resize this one down to 1k to match it.

Comment (fgsch) [reply internal/abi/fastly/hostcalls_guest.go:3414]: Thank you for the explanation. I think it's fine. I was more concerned about being smaller, in which cases we would need to do 2 calls. I think we can leave it as is.

File: integration_tests/device_detection/fastly.toml (hunk 2 above)
Comment (fgsch) [reply integration_tests/device_detection/fastly.toml:19]: Not a blocker, btw.

**Review by fgsch — APPROVED** [review]
Thank you!

File: device/device_detection.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	}
+
+	if err := json.Unmarshal(raw, &d.info); err != nil {
+		return d, err
+	}
+
+	return d, nil
+}
+
+// Name returns the name of the client device.
```
[the remark this answers is not in the document]
Comment (joeshaw) [reply device/device_detection.go:66]: The device stuff is the only stuff we're committing to at the moment.  I force-pushed a commit that removes the user agent and OS stuff for now.

**Review by fgsch — APPROVED** [review]
Still 👍

**Pull request merged**